# Odds, Log Odds, and Decision Boundary (Complete Notes)

Before understanding Logistic Regression completely, we need to understand three important concepts:

1. Odds
2. Log Odds (Logit)
3. Decision Boundary

These concepts explain **why Logistic Regression uses the sigmoid function** and **how it makes predictions**.

---

# 1. Odds

Probability tells us:

> **How likely is an event to happen?**

Odds tell us:

> **How many times more likely is the event to happen than not happen?**

## Formula

$$
\text{Odds}
=
\frac{P}{1-P}
$$

where

- $P$ = Probability of the positive class

---

## Example 1

Suppose

$$
P = 0.5
$$

Then,

$$
\text{Odds}
=
\frac{0.5}{1-0.5}
=
\frac{0.5}{0.5}
=
1
$$

Interpretation:

The event is **equally likely to happen as not happen**.

Odds = **1 : 1**

---

## Example 2

Suppose

$$
P = 0.8
$$

Then,

$$
\text{Odds}
=
\frac{0.8}{0.2}
=
4
$$

Interpretation:

The event is **4 times more likely to happen than not happen**.

Odds = **4 : 1**

---

## Example 3

Suppose

$$
P = 0.2
$$

Then,

$$
\text{Odds}
=
\frac{0.2}{0.8}
=
0.25
$$

Interpretation:

The event is only **one-fourth as likely to happen as not happen**.

Odds = **1 : 4**

---

## Understanding Odds

| Probability ($P$) | Odds | Interpretation |
|------------------:|------|----------------|
| 0.1 | 0.11 | Very unlikely |
| 0.2 | 0.25 | Unlikely |
| 0.5 | 1 | Equally likely |
| 0.8 | 4 | Very likely |
| 0.9 | 9 | Extremely likely |

---

## Problem with Odds

Notice how the odds range is:

$$
0
\rightarrow
\infty
$$

Examples:

| Probability | Odds |
|-------------|------|
|0|0|
|0.2|0.25|
|0.5|1|
|0.8|4|
|0.99|99|

The values become extremely large.

This is inconvenient for building a linear model.

To solve this problem, we take the logarithm of the odds.

---

# 2. Log Odds (Logit)

Log Odds are simply the logarithm of the odds.

They are also called the **Logit**.

## Formula

$$
\text{Log Odds}
=
\log\left(
\frac{P}{1-P}
\right)
$$

---

## Example 1

Suppose

$$
P = 0.5
$$

Odds

$$
=
1
$$

Log Odds

$$
=
\log(1)
=
0
$$

---

## Example 2

Suppose

$$
P = 0.8
$$

Odds

$$
=
4
$$

Log Odds

$$
=
\log(4)
=
1.386
$$

---

## Example 3

Suppose

$$
P = 0.2
$$

Odds

$$
=
0.25
$$

Log Odds

$$
=
\log(0.25)
=
-1.386
$$

---

## Understanding Log Odds

| Probability | Odds | Log Odds |
|-------------|------|----------|
|0.1|0.11|-2.197|
|0.2|0.25|-1.386|
|0.5|1|0|
|0.8|4|1.386|
|0.9|9|2.197|

Notice something interesting.

- Probability less than **0.5** → Negative Log Odds
- Probability equal to **0.5** → Zero Log Odds
- Probability greater than **0.5** → Positive Log Odds

---

## Why Use Log Odds?

The probability is restricted to

$$
0 \le P \le 1
$$

A linear equation

$$
y = wx + b
$$

can produce values like

$$
-50,\;10,\;250,\;-1000
$$

Clearly, these are **not valid probabilities**.

Instead of predicting probabilities directly, Logistic Regression predicts the **Log Odds**.

The beautiful idea behind Logistic Regression is

$$
\boxed{
\log\left(\frac{P}{1-P}\right)
=
wx+b
}
$$

This means that the **Log Odds** can take **any real value**, making them perfect for a linear equation.

---

## From Log Odds to Probability

Starting with

$$
\log\left(\frac{P}{1-P}\right)
=
wx+b
$$

Exponentiating both sides,

$$
\frac{P}{1-P}
=
e^{wx+b}
$$

Rearranging,

$$
P
=
\frac{e^{wx+b}}
{1+e^{wx+b}}
$$

Multiplying numerator and denominator by $e^{-(wx+b)}$,

$$
P
=
\frac{1}
{1+e^{-(wx+b)}}
$$

This is exactly the **Sigmoid Function**.

Therefore,

$$
\boxed{
P
=
\frac{1}
{1+e^{-(wx+b)}}
}
$$

So the sigmoid function is **not chosen arbitrarily**.

It is obtained by solving the Log Odds equation for the probability.

---

# 3. Decision Boundary

The model predicts a probability.

Example:

| Student | Probability of Passing |
|----------|-----------------------:|
|A|0.95|
|B|0.82|
|C|0.63|
|D|0.41|
|E|0.12|

Eventually, we must convert these probabilities into classes.

---

## Threshold

The default threshold is

$$
0.5
$$

Decision rule:

```text
If Probability ≥ 0.5
    Predict Class = 1

Else
    Predict Class = 0
```

---

## Why 0.5?

When

$$
P = 0.5
$$

the model is equally confident about both classes.

This naturally becomes the dividing line.

---

## Deriving the Decision Boundary

The sigmoid function is

$$
P
=
\frac{1}
{1+e^{-z}}
$$

where

$$
z
=
wx+b
$$

At the decision boundary,

$$
P
=
0.5
$$

Substituting,

$$
0.5
=
\frac{1}
{1+e^{-z}}
$$

Multiplying both sides,

$$
0.5(1+e^{-z})
=
1
$$

$$
1+e^{-z}
=
2
$$

$$
e^{-z}
=
1
$$

Taking the natural logarithm,

$$
-z
=
\log(1)
=
0
$$

Therefore,

$$
z
=
0
$$

Since

$$
z
=
wx+b
$$

the decision boundary is

$$
\boxed{
wx+b=0
}
$$

---

## One Feature Example

Suppose

$$
w=2
$$

and

$$
b=-6
$$

Decision Boundary

$$
2x-6=0
$$

Solving,

$$
x=3
$$

Interpretation:

- If $x > 3$, predict **Class 1**
- If $x < 3$, predict **Class 0**

---

## Two Features

Suppose

$$
2x_1+3x_2-6=0
$$

This equation represents a **straight line**.

The line separates the two classes.

Everything on one side belongs to one class.

Everything on the other side belongs to the other class.

---

## Three Features

Suppose

$$
w_1x_1+w_2x_2+w_3x_3+b=0
$$

This is no longer a line.

It is called a **hyperplane**.

Logistic Regression always tries to find the hyperplane that best separates the classes.

---

# Summary

| Concept | Formula | Meaning |
|---------|---------|---------|
| **Probability** | $P$ | Chance of the positive class |
| **Odds** | $\frac{P}{1-P}$ | How many times more likely the event is to happen than not happen |
| **Log Odds (Logit)** | $\log\left(\frac{P}{1-P}\right)$ | Linear representation of the odds |
| **Logistic Regression Model** | $\log\left(\frac{P}{1-P}\right)=wx+b$ | Assumes the log odds are a linear function of the features |
| **Sigmoid Function** | $P=\frac{1}{1+e^{-(wx+b)}}$ | Converts the linear model into a probability |
| **Decision Boundary** | $wx+b=0$ | Boundary that separates the two classes |

---

# Key Takeaways

- **Probability** tells us how likely an event is.
- **Odds** compare the probability of success to the probability of failure.
- **Log Odds** transform the odds into values that range from $-\infty$ to $+\infty$, making them suitable for linear models.
- Logistic Regression assumes that the **Log Odds** are a linear combination of the input features.
- Solving the Log Odds equation gives us the **Sigmoid Function**.
- The **Decision Boundary** occurs where the predicted probability is **0.5**, which simplifies to the equation:

$$
wx+b=0
$$

This boundary is a **point** in one dimension, a **line** in two dimensions, and a **hyperplane** in higher dimensions.

We'll use **gradient descent** because there is no closed-form solution.

We'll do **one complete iteration** by hand.

---

# Problem

Suppose we have this dataset:

| Hours Studied ($x$) | Passed ($y$) |
|---------------------|--------------|
| 1 | 0 |
| 2 | 0 |
| 3 | 1 |

We want to learn a model that predicts whether a student passes.

---

# Step 1: Initialize the Parameters

Let's start with:

```python
w = 0
b = 0
learning_rate = 0.1
```

---

# Step 2: Compute $z$

The linear part of logistic regression is

$$
z = wx + b
$$

For each student:

| $x$ | Calculation | $z$ |
|-----|-------------|-----|
| 1 | $0 \times 1 + 0$ | 0 |
| 2 | $0 \times 2 + 0$ | 0 |
| 3 | $0 \times 3 + 0$ | 0 |

---

# Step 3: Apply the Sigmoid Function

The prediction is given by the sigmoid function:

$$
p = \sigma(z) = \frac{1}{1 + e^{-z}}
$$

For $z = 0$,

$$
p = \frac{1}{1 + e^0}
= \frac{1}{2}
= 0.5
$$

So every prediction is:

| $x$ | Actual ($y$) | Prediction ($p$) |
|-----|--------------|------------------|
| 1 | 0 | 0.5 |
| 2 | 0 | 0.5 |
| 3 | 1 | 0.5 |

---

# Step 4: Compute the Cross-Entropy Loss

The loss for one sample is

$$
L = -\left(y \log(p) + (1-y)\log(1-p)\right)
$$

Let's compute the loss for each sample.

---

## Sample 1

Actual = 0

Prediction = 0.5

$$
L_1
=
-\left(0\log(0.5)+1\log(0.5)\right)
$$

$$
= -\log(0.5)
$$

$$
= 0.6931
$$

---

## Sample 2

Exactly the same.

$$
L_2 = 0.6931
$$

---

## Sample 3

Actual = 1

Prediction = 0.5

$$
L_3
=
-\left(1\log(0.5)\right)
$$

$$
= 0.6931
$$

---

The average loss is

$$
\frac{0.6931 + 0.6931 + 0.6931}{3}
=
0.6931
$$

This is our starting error.

---

# Step 5: Compute the Gradients

After differentiating the cost function (using the chain rule and the derivative of the sigmoid), the gradients simplify to

$$
\frac{\partial J}{\partial w}
=
\frac{1}{m}
\sum_{i=1}^{m}
(p_i-y_i)x_i
$$

and

$$
\frac{\partial J}{\partial b}
=
\frac{1}{m}
\sum_{i=1}^{m}
(p_i-y_i)
$$

Let's compute them.

---

## Gradient for $w$

| $x$ | $y$ | $p$ | $p-y$ | $(p-y)x$ |
|-----|-----|-----|--------|----------|
| 1 | 0 | 0.5 | 0.5 | 0.5 |
| 2 | 0 | 0.5 | 0.5 | 1.0 |
| 3 | 1 | 0.5 | -0.5 | -1.5 |

The sum is

$$
0.5 + 1 - 1.5 = 0
$$

Therefore,

$$
\frac{\partial J}{\partial w}
=
0
$$

Interesting! The positive and negative contributions cancel out.

---

## Gradient for $b$

| $p-y$ |
|--------|
| 0.5 |
| 0.5 |
| -0.5 |

The sum is

$$
0.5
$$

The average is

$$
\frac{0.5}{3}
=
0.1667
$$

Therefore,

$$
\frac{\partial J}{\partial b}
=
0.1667
$$

---

# Step 6: Update the Parameters

Gradient descent updates the parameters as follows:

$$
w
=
w
-
\alpha
\frac{\partial J}{\partial w}
$$

$$
b
=
b
-
\alpha
\frac{\partial J}{\partial b}
$$

where the learning rate is

$$
\alpha = 0.1
$$

### Update $w$

$$
w
=
0
-
0.1(0)
=
0
$$

### Update $b$

$$
b
=
0
-
0.1(0.1667)
=
-0.01667
$$

The new parameters are

```python
w = 0
b = -0.01667
```

---

# Step 7: Predict Again

Now,

$$
z = wx + b
$$

Since $w = 0$,

$$
z = -0.01667
$$

for every sample.

The new prediction becomes

$$
\sigma(-0.01667)
\approx
0.4958
$$

So every prediction shifts slightly from **0.5000** to **0.4958**, moving in the direction that reduces the loss for the two negative examples.

---

# Where Did Those Gradient Formulas Come From?

The overall cost function is

$$
J
=
-\frac{1}{m}
\sum_{i=1}^{m}
\left[
y_i\log(p_i)
+
(1-y_i)\log(1-p_i)
\right]
$$

where

$$
p = \sigma(z)
$$

and

$$
z = wx + b
$$

To compute the gradient with respect to $w$, we use the **chain rule**:

$$
\frac{\partial J}{\partial w}
=
\frac{\partial J}{\partial p}
\cdot
\frac{\partial p}{\partial z}
\cdot
\frac{\partial z}{\partial w}
$$

The three derivatives are:

- $\frac{\partial J}{\partial p}$ → derivative of the cross-entropy loss
- $\frac{\partial p}{\partial z} = p(1-p)$ → derivative of the sigmoid function
- $\frac{\partial z}{\partial w} = x$

After simplification, the $p(1-p)$ terms cancel beautifully, giving the elegant result

$$
\boxed{
\frac{\partial J}{\partial w}
=
\frac{1}{m}
\sum_{i=1}^{m}
(p_i-y_i)x_i
}
$$

Similarly,

$$
\boxed{
\frac{\partial J}{\partial b}
=
\frac{1}{m}
\sum_{i=1}^{m}
(p_i-y_i)
}
$$

These are exactly the gradients used by libraries like **scikit-learn**, **TensorFlow**, and **PyTorch** when training logistic regression using gradient descent or its variants.

---

# A Better Example for Teaching

This dataset happened to give

$$
\frac{\partial J}{\partial w}=0
$$

which is mathematically correct but not very illustrative.

A better teaching dataset is:

| $x$ | $y$ |
|-----|-----|
| 1 | 0 |
| 2 | 0 |
| 4 | 1 |
| 5 | 1 |

For this dataset, **both** gradients are non-zero, so students can clearly see **both** $w$ and $b$ changing after the first iteration, making it much easier to understand how logistic regression learns.

In [1]:
import numpy as np
import pandas as pd


class LogisticRegressionScratch:

    def __init__(self, learning_rate=0.1, epochs=1000):
        self.lr = learning_rate
        self.epochs = epochs
        self.w = 0
        self.b = 0

    # -------------------------
    # Sigmoid Function
    # -------------------------
    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))

    # -------------------------
    # Train the model
    # -------------------------
    def fit(self, df):

        X = df["x"].values
        y = df["y"].values

        m = len(X)

        # Initialize parameters
        self.w = 0
        self.b = 0

        for epoch in range(self.epochs):

            # Step 1
            z = self.w * X + self.b

            # Step 2
            p = self.sigmoid(z)

            # Step 3 (Gradients)

            dw = (1/m) * np.sum((p - y) * X)

            db = (1/m) * np.sum(p - y)

            # Step 4 (Gradient Descent)

            self.w -= self.lr * dw
            self.b -= self.lr * db

            # Print every 100 epochs
            if epoch % 100 == 0:

                loss = -(1/m) * np.sum(
                    y*np.log(p + 1e-15)
                    +
                    (1-y)*np.log(1-p + 1e-15)
                )

                print(
                    f"Epoch {epoch:4d} "
                    f"Loss={loss:.5f} "
                    f"w={self.w:.4f} "
                    f"b={self.b:.4f}"
                )

    # -------------------------
    # Probability Prediction
    # -------------------------
    def predict_proba(self, X):

        X = np.array(X)

        return self.sigmoid(self.w * X + self.b)

    # -------------------------
    # Class Prediction
    # -------------------------
    def predict(self, X):

        probs = self.predict_proba(X)

        return (probs >= 0.5).astype(int)

In [2]:
df = pd.DataFrame({

    "x":[1,2,3,4,5,6,7,8],

    "y":[0,0,0,0,1,1,1,1]

})

model = LogisticRegressionScratch(
            learning_rate=0.1,
            epochs=2000)

model.fit(df)

Epoch    0 Loss=0.69315 w=0.1000 b=0.0000
Epoch  100 Loss=0.40807 w=0.4334 b=-1.3891
Epoch  200 Loss=0.31438 w=0.6222 b=-2.3311
Epoch  300 Loss=0.26310 w=0.7651 b=-3.0294
Epoch  400 Loss=0.23065 w=0.8804 b=-3.5854
Epoch  500 Loss=0.20803 w=0.9775 b=-4.0498
Epoch  600 Loss=0.19120 w=1.0618 b=-4.4506
Epoch  700 Loss=0.17808 w=1.1366 b=-4.8045
Epoch  800 Loss=0.16748 w=1.2042 b=-5.1226
Epoch  900 Loss=0.15869 w=1.2659 b=-5.4123
Epoch 1000 Loss=0.15124 w=1.3229 b=-5.6791
Epoch 1100 Loss=0.14481 w=1.3760 b=-5.9267
Epoch 1200 Loss=0.13920 w=1.4257 b=-6.1582
Epoch 1300 Loss=0.13423 w=1.4725 b=-6.3760
Epoch 1400 Loss=0.12979 w=1.5169 b=-6.5818
Epoch 1500 Loss=0.12579 w=1.5591 b=-6.7772
Epoch 1600 Loss=0.12216 w=1.5993 b=-6.9634
Epoch 1700 Loss=0.11884 w=1.6378 b=-7.1414
Epoch 1800 Loss=0.11579 w=1.6748 b=-7.3120
Epoch 1900 Loss=0.11298 w=1.7103 b=-7.4759


In [3]:
model.predict_proba([1,3,4,5,6])

array([0.00276476, 0.0832118 , 0.3418156 , 0.74820465, 0.94445075])


# Classification Evaluation Metrics (Complete Notes)

After building a classification model (Logistic Regression, KNN, Decision Tree, Random Forest, etc.), the next question is:

> **How good is the model?**

Unlike regression, we cannot use metrics like MAE, MSE, or RMSE.

Instead, we use **classification metrics**.

---

# 1. Confusion Matrix (Most Important)

Everything else is derived from the **Confusion Matrix**.

Imagine we built a model to predict whether a student passes.

| Actual | Predicted |
|---------|-----------|
| Pass | Pass |
| Fail | Pass |
| Pass | Fail |
| Fail | Fail |

These four possibilities form the confusion matrix.

|                     | **Predicted Positive** | **Predicted Negative** |
|---------------------|------------------------|------------------------|
| **Actual Positive** | **True Positive (TP)** | **False Negative (FN)** |
| **Actual Negative** | **False Positive (FP)** | **True Negative (TN)** |

The confusion matrix is the **foundation of all classification metrics**.

---

## True Positive (TP)

The model predicted **Positive**, and the actual class was also **Positive**.

Example:

| Actual | Prediction |
|---------|------------|
| Passed | Passed |

Correct prediction.

---

## True Negative (TN)

The model predicted **Negative**, and the actual class was also **Negative**.

Example:

| Actual | Prediction |
|---------|------------|
| Failed | Failed |

Correct prediction.

---

## False Positive (FP)

The model predicted **Positive**, but the actual class was **Negative**.

Also called:

> **Type I Error**

Example:

| Actual | Prediction |
|---------|------------|
| Failed | Passed |

The student failed, but the model predicted they passed.

---

## False Negative (FN)

The model predicted **Negative**, but the actual class was **Positive**.

Also called:

> **Type II Error**

Example:

| Actual | Prediction |
|---------|------------|
| Passed | Failed |

The student actually passed, but the model predicted failure.

---

# Example Confusion Matrix

Suppose we tested our model on **100 students**.

|                 | Predicted Pass | Predicted Fail |
|-----------------|---------------:|---------------:|
| Actual Pass     | 40             | 10             |
| Actual Fail     | 5              | 45             |

Therefore,

```text
TP = 40
FP = 5
FN = 10
TN = 45
```

These four values are enough to compute every important classification metric.

---

# 2. Accuracy

Accuracy answers the question:

> **Out of all predictions, how many were correct?**

## Formula

$$
\text{Accuracy} =
\frac{TP + TN}
{TP + TN + FP + FN}
$$

Using our example,

$$
\text{Accuracy}
=
\frac{40 + 45}{100}
=
0.85
$$

Accuracy = **85%**

---

## When Accuracy Works Well

Accuracy works well when the classes are **balanced**.

Example:

- 50 Cats
- 50 Dogs

---

## When Accuracy Fails

Suppose we have **1000 emails**.

- 990 Normal
- 10 Spam

A model predicts:

```text
Everything = Normal
```

Accuracy becomes

$$
\frac{990}{1000}
=
99\%
$$

Amazing?

**No.**

The model failed to detect **every spam email**.

This is why we need Precision and Recall.

---

# 3. Precision

Precision answers:

> **Out of everything the model predicted as Positive, how many were actually Positive?**

## Formula

$$
\text{Precision}
=
\frac{TP}
{TP + FP}
$$

Using our example,

$$
\text{Precision}
=
\frac{40}
{40 + 5}
=
0.889
$$

Precision = **88.9%**

---

## Precision Example

Suppose we are detecting cancer.

Whenever the model predicts:

```text
Cancer
```

Precision asks:

> **How often is that prediction correct?**

High Precision means:

- Few False Positives
- Very reliable positive predictions

---

# 4. Recall (Sensitivity)

Recall answers:

> **Out of all actual Positive cases, how many did we correctly identify?**

## Formula

$$
\text{Recall}
=
\frac{TP}
{TP + FN}
$$

Using our example,

$$
\text{Recall}
=
\frac{40}
{40 + 10}
=
0.80
$$

Recall = **80%**

---

## Recall Example

Suppose:

- 100 patients actually have cancer.

The model correctly detects:

- 80 patients

Then,

$$
\text{Recall}
=
\frac{80}{100}
=
80\%
$$

The remaining 20 patients are **False Negatives**.

---

# Precision vs Recall

Imagine an airport security scanner.

## High Precision

Whenever the scanner raises an alarm,

it is almost always correct.

Very few innocent passengers are stopped.

---

## High Recall

Almost every dangerous item is detected,

even if many innocent passengers are searched.

---

## Rule of Thumb

Medical diagnosis

→ High Recall

(Do not miss sick patients.)

Spam filtering

→ High Precision

(Do not classify important emails as spam.)

---

# 5. F1 Score

Sometimes,

- Precision is high.
- Recall is low.

Which model is better?

The **F1 Score** combines Precision and Recall into a single number.

## Formula

$$
F_1
=
2
\times
\frac{\text{Precision}
\times
\text{Recall}}
{\text{Precision}
+
\text{Recall}}
$$

Using

$$
\text{Precision}
=
0.889
$$

and

$$
\text{Recall}
=
0.80
$$

we get

$$
F_1
=
2
\times
\frac{0.889 \times 0.80}
{0.889 + 0.80}
=
0.842
$$

F1 Score = **84.2%**

---

## Why Harmonic Mean?

Suppose

- Precision = 100%
- Recall = 10%

The arithmetic mean is

$$
\frac{100 + 10}{2}
=
55\%
$$

which seems acceptable.

However, the model misses **90% of all positive cases**.

The harmonic mean penalizes this imbalance, ensuring that **both Precision and Recall must be high** for the F1 Score to be high.

---

# Relationship Between Metrics

Notice that:

- Accuracy uses **all four values**.
- Precision focuses on **predicted positives**.
- Recall focuses on **actual positives**.
- F1 balances Precision and Recall.

---

# 6. ROC Curve

Most classification models predict a **probability**, not just a class.

Example:

| Student | Probability of Passing |
|----------|-----------------------:|
| A | 0.98 |
| B | 0.91 |
| C | 0.80 |
| D | 0.45 |
| E | 0.10 |

Normally,

```text
Probability ≥ 0.5  → Positive
Probability < 0.5  → Negative
```

But what if we change the threshold?

- 0.9
- 0.7
- 0.5
- 0.3

The confusion matrix changes every time.

The **ROC Curve** summarizes the model's performance across **all possible thresholds**.

---

## Axes of the ROC Curve

### X-axis

False Positive Rate (FPR)

$$
FPR
=
\frac{FP}
{FP + TN}
$$

### Y-axis

True Positive Rate (TPR)

$$
TPR
=
\frac{TP}
{TP + FN}
$$

Notice that

$$
TPR
=
\text{Recall}
$$

Each point on the ROC curve corresponds to a different classification threshold.

---

## Ideal ROC Curve

The closer the curve is to the **top-left corner**, the better.

A random classifier follows the diagonal line.

```text
TPR

1.0 |■■■■■■■
    |      ■
    |       ■
    |        ■
0.0 +--------------------> FPR
     0                 1
```

---

# 7. AUC (Area Under the Curve)

AUC is simply the **Area Under the ROC Curve**.

It summarizes the overall performance of a classifier.

## Interpretation

| AUC | Interpretation |
|------|----------------|
| 1.0 | Perfect classifier |
| 0.9 – 1.0 | Excellent |
| 0.8 – 0.9 | Good |
| 0.7 – 0.8 | Fair |
| 0.6 – 0.7 | Poor |
| 0.5 | Random guessing |

An AUC of **0.5** means the classifier performs no better than random guessing.

The closer the AUC is to **1**, the better the classifier separates the two classes.

---

# Summary Table

| Metric | Formula | Best Used When |
|---------|---------|----------------|
| **Confusion Matrix** | TP, TN, FP, FN | Foundation of all classification metrics |
| **Accuracy** | $\frac{TP+TN}{TP+TN+FP+FN}$ | Balanced datasets |
| **Precision** | $\frac{TP}{TP+FP}$ | False Positives are costly |
| **Recall** | $\frac{TP}{TP+FN}$ | False Negatives are costly |
| **F1 Score** | $2\times\frac{\text{Precision}\times\text{Recall}}{\text{Precision}+\text{Recall}}$ | Need a balance between Precision and Recall |
| **ROC Curve** | TPR vs FPR | Compare performance across thresholds |
| **AUC** | Area Under ROC Curve | Compare classifiers using a single score |

---

# Interview Questions

1. What is a Confusion Matrix?
2. Explain TP, TN, FP, and FN with examples.
3. Why can Accuracy be misleading for imbalanced datasets?
4. What is the difference between Precision and Recall?
5. In which situations is Recall more important than Precision?
6. Why does the F1 Score use the harmonic mean instead of the arithmetic mean?
7. What does the ROC Curve represent?
8. What is the difference between ROC and AUC?
9. What does an AUC of 0.5 indicate?
10. Can a model have high Accuracy but low Recall? Explain with an example.

---

# Key Takeaways

- The **Confusion Matrix** is the foundation of all classification evaluation metrics.
- **Accuracy** measures overall correctness.
- **Precision** measures the reliability of positive predictions.
- **Recall** measures the ability to detect all positive cases.
- **F1 Score** balances Precision and Recall.
- **ROC Curve** evaluates model performance across different probability thresholds.
- **AUC** summarizes the ROC Curve into a single number.

These metrics are applicable to almost every classification algorithm, including:

- Logistic Regression
- K-Nearest Neighbors (KNN)
- Decision Trees
- Random Forest
- Support Vector Machines (SVM)
- Naive Bayes
- Gradient Boosting
- XGBoost

